dom.bs

```diff
@@ -4986,6 +5033,27 @@ within <a>this</a>.
 <p>The <dfn method for=Node><code>removeChild(<var>child</var>)</code></dfn> method steps are to
 return the result of <a>pre-removing</a> <var>child</var> from <a>this</a>.
 
+<p>The <dfn method for=Node><code>moveBefore(<var>node</var>, <var>child</var>)</code></dfn>
+method steps are:
+
+<ol>
+ <li><p>Let <var>perform state-preserving atomic move</var> be true if <a>this</a> is
+ <a>connected</a>, <var>node</var> is <a>connected</a>, <a>this</a>'s <a for=Node>node
+ document</a> is <a>fully active</a>, and <a>this</a>'s <a for=/>root</a> is the same as
+ <var>node</var>'s <a for=/>root</a>.</p></li>.
+
+ <li><p>If <var>perform state-preserving atomic move</var> is true, then set <a>this</a>'s <a
+ for=Node>node document</a>'s <a>state-preserving atomic move in progress</a> to true.</p></li>
+
+ <li><p>Let <var>return node</var> be the result of <a>pre-inserting</a> <var>node</var> into
+ <a>this</a> before <var>child</var>.</p></li>
```

Comment:
Yeah I think initially I saw it as a separate primitive too, but the more I looked at it, the more the difference between the two seemed really subtle. It mostly has to do with MutationObservers (and half of the relevant logic here is tucked away in the "remove" primitive) and *not* running the post-connection steps. So I feel like we'd end up with a near line-by-line copy of "insert", modulo one or two small differences. I'll take another look to see if my intuition is accurate, but I do kinda suspect this is where we'd end up.

Reply:
It'd be good to have more details. Because you can't move a `DocumentFragment` so a vast number of checks don't apply. Adopt won't ever run.

There might be a number of range and shadow tree checks that end we end up duplicating, but perhaps that calls for abstracting those instead. At least to me a state flag seems rather unappealing.

Reply:
It's possible there is enough from "insert" that we'd omit in "move", to make it creation worth it, sure.

> At least to me a state flag seems rather unappealing.

With a separate "move" primitive _and_ the decision to throw when we can't "move", we could get rid of the state variable that's currently in "insert" (i.e., `<var>statePreservingAtomicMoveInProgress</var>`). But the state bool on _`Document`_ is what other specifications will refer to in their removal steps to react to a move appropriately, so I'm not sure we can get rid of _that_ one.

Reply:
Hmm, wouldn't we give other specifications "move steps"?

Reply:
I thought about this a bit more. I think we're happy to create a new "move" primitive algorithm following the "insert" and "remove" algorithms. As you mentioned, it will be a simpler algorithm than "insert" since we'll avoid the `DocumentFragment` stuff — oh and also the post-connection stuff, among maybe more things.

There's still the discussion around introducing either (**a**) new "move" steps/hook that other specs provide to react to moves, vs (**b**) a Document-scoped boolean that the new "move" algorithm sets, for existing insertion/removal steps to use, to not reset certain state inside those steps.

Option (b) is the direction of the current PR, which we think makes sense after implementing this in Chromium. I think this is because so far we've found no behavior that we specifically want to do during a move that isn't done by insertion/removal — so any "move steps" we provided would just be empty. **Does that make sense?**

Now, I think the real question here is whether we want to:

1. Keep running the [insertion](https://dom.spec.whatwg.org/#concept-node-insert-ext)/[removal](https://dom.spec.whatwg.org/#concept-node-remove-ext) steps during a "move" operation, and just pepper `if during atomic move is true, don't run some of these substeps!` wherever state would ordinarily be reset
    - Cancel out [the iframe removal steps](https://html.spec.whatwg.org/C#the-iframe-element:html-element-insertion-steps)
    - Cancel [img relevant mutations](https://html.spec.whatwg.org/C#reacting-to-dom-mutations:html-element-insertion-steps), to avoid re-firing load events
1. Simply not run insertion/removal hooks _at all_ during a move. This is only feasible if we determine that there are absolutely no steps in any spec's existing insertion/removal hooks that need to run during a move. We'll have to do an audit to determine this, which I'm starting now.

So the way I see it, those are probably our options, as opposed to creating new move steps which I think from our findings so far, would be unused. (We'll course-correct if this finding changes).